cmd/snap-bootstrap/blkid/blkid.go

```diff
@@ -0,0 +1,171 @@
+	Close()
+	// EnablePartitions is a wrapper for blkid_probe_enable_partitions
+	EnablePartitions(value bool)
+	// EnableSuperblocks is a wrapper for blkid_probe_enable_superblocks
+	EnableSuperblocks(value bool)
+	// SetPartitionsFlags is a wrapper for blkid_probe_set_partitions_flags
+	SetPartitionsFlags(flags int)
+	// DoSafeprobe is a wrapper for blkid_do_safeprobe
+	DoSafeprobe() error
+	// GetPartitions is a wrapper for blkid_probe_get_partitions
+	GetPartitions() (AbstractBlkidPartlist, error)
+}
+
+// AbstractBlkidPartlist is a wrapper for blkid_partlist
+type AbstractBlkidPartlist interface {
+	// GetPartitions is a wrapper for blkid_partlist_get_partition
+	// and blkid_partlist_numof_partitions.
+	GetPartitions() []AbstractBlkidPartition
+}
+
+// AbstractBlkidPartition is a wrapper for blkid_partition
+type AbstractBlkidPartition interface {
+	// GetName is a wrapper for blkid_partition_get_name
+	GetName() string
+	// GetUUID is a wrapper for blkid_partition_get_uuid
+	GetUUID() string
+}
+
+type blkidProbe struct {
+	probeHandle C.blkid_probe
+}
+
+func newProbeFromFilenameImpl(node string) (AbstractBlkidProbe, error) {
+	cnode := C.CString(node)
+	defer C.free(unsafe.Pointer(cnode))
+	C.blkid_new_probe_from_filename(cnode)
+	probe, err := C.blkid_new_probe_from_filename(cnode)
+	if probe == nil {
+		return nil, err
+	}
+	return &blkidProbe{probe}, nil
+}
+
+var NewProbeFromFilename = newProbeFromFilenameImpl
+
+func (p *blkidProbe) LookupValue(entryName string) (string, error) {
+	var value *C.char
+	var value_len C.size_t
+	cname := C.CString(entryName)
+	defer C.free(unsafe.Pointer(cname))
+	res := C.blkid_probe_lookup_value(p.probeHandle, cname, &value, &value_len)
+	if res < 0 {
+		return "", fmt.Errorf("Probe value was not found: %s", entryName)
+	}
+	if value_len > 0 {
+		return C.GoStringN(value, C.int(value_len-1)), nil
+	} else {
+		return "", fmt.Errorf("Probe value has unexpected size")
+	}
+}
+
+func (p *blkidProbe) Close() {
+	C.blkid_free_probe(p.probeHandle)
+}
+
+func (p *blkidProbe) EnablePartitions(value bool) {
+	v := 0
+	if value {
+		v = 1
+	}
+	C.blkid_probe_enable_partitions(p.probeHandle, C.int(v))
+}
+
+func (p *blkidProbe) EnableSuperblocks(value bool) {
+	v := 0
+	if value {
+		v = 1
+	}
+	C.blkid_probe_enable_superblocks(p.probeHandle, C.int(v))
+}
+
+func (p *blkidProbe) SetPartitionsFlags(flags int) {
+	C.blkid_probe_set_partitions_flags(p.probeHandle, C.int(flags))
+}
+
+func (p *blkidProbe) DoSafeprobe() error {
+	res, err := C.blkid_do_safeprobe(p.probeHandle)
+	if res < 0 {
+		return err
+	}
+	return nil
+}
+
+type blkidPartlist struct {
+	partlistHandle C.blkid_partlist
+}
+
+func (p *blkidProbe) GetPartitions() (AbstractBlkidPartlist, error) {
+	partitions, err := C.blkid_probe_get_partitions(p.probeHandle)
+	if partitions == nil {
+		return nil, err
+	}
+	return &blkidPartlist{partitions}, nil
+}
+
+type blkidPartition struct {
+	partitionHandle C.blkid_partition
```

Comment:
is this deallocated when the probe is?

Reply:
That is my understanding from the documentation. But it is all clear. There is no other free than `blkid_free_probe`.

> Deallocates the probe struct, buffers and all allocated data that are associated with this probing control struct.